### PR TITLE
implement ReadableStream Default Tee

### PR DESCRIPTION
**Pull request by Taym95**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/script/dom/underlyingsourcecontainer.rs [hunk 1]
```diff
@@ -129,10 +151,65 @@ impl UnderlyingSourceContainer {
...
+                    stream: *stream,
+                    branch1: *branch1,
+                    branch2: *branch2,
+                    reading: reading.clone(),
+                    read_again: read_again.clone(),
+                    canceled1: canceled1.clone(),
+                    canceled2: canceled2.clone(),
+                    clon_for_branch2: clon_for_branch2.clone(),
+                    cancel_promise: cancel_promise.clone(),
+                    global: self.global(),
```
Comment (Taym95) [inline components/script/dom/underlyingsourcecontainer.rs]: `GlobalScope ` is needed inside `ReadRequest ::chunk_steps` for `structuredclone`, this is not working, how can I pass `GlobalScope` here?

File: components/script/dom/readablestream.rs [hunk 2]
```diff
@@ -591,6 +607,73 @@ impl ReadableStream {
...
+        let branch1 = MutNullableDom::new(None);
+        // // Let branch2 be undefined.
+        let branch2 = MutNullableDom::new(None);
+
+        // Let cancelPromise be a new promise.
+        let cancel_promise = Promise::new(&self.reflector_.global(), CanGc::note());
+
+        let underlying_sourc_type = UnderlyingSourceType::Tee {
+            reader,
+            stream: MutNullableDom::new(Some(self)),
```
Comment (Taym95) [inline components/script/dom/readablestream.rs]: I need to pass `stream, branch1, branch2' because I need to call close and error functions

Comment (Taym95) [reply components/script/dom/readablestream.rs]: I tried to use closures but I ended up with lifetime issues

File: components/script/dom/underlyingsourcecontainer.rs (hunk 1 above)
Comment (wusyong) [reply components/script/dom/underlyingsourcecontainer.rs]: The stream is a DOM object. could it call `global()` from it?

Comment (Taym95) [reply components/script/dom/underlyingsourcecontainer.rs]: Ah yes indeed

File: components/script/dom/underlyingsourcecontainer.rs [hunk 3]
```diff
@@ -34,6 +38,23 @@ pub enum UnderlyingSourceType {
     FetchResponse,
     /// A JS object as underlying source.
     Js(JsUnderlyingSource),
+    /// Tee
+    Tee {
```
Comment (gterzian) [inline components/script/dom/underlyingsourcecontainer.rs:43]: Here I think we can do `Tee(TeeUnderlyingSource)`, and put the data in a `struct TeeUnderlyingSource`.

File: components/script/dom/underlyingsourcecontainer.rs [hunk 4]
```diff
@@ -34,6 +38,23 @@ pub enum UnderlyingSourceType {
...
+        reader: DomRoot<ReadableStreamDefaultReader>,
```
Comment (gterzian) [inline components/script/dom/underlyingsourcecontainer.rs]: This should be `Dom`

File: components/script/dom/underlyingsourcecontainer.rs [hunk 5]
```diff
@@ -34,6 +38,23 @@ pub enum UnderlyingSourceType {
...
+        stream: MutNullableDom<ReadableStream>,
+        branch1: MutNullableDom<ReadableStream>,
+        branch2: MutNullableDom<ReadableStream>,
```
Comment (gterzian) [inline components/script/dom/underlyingsourcecontainer.rs]: Do you need them to be mutable? If not, use `Dom` as well.

File: components/script/dom/readablestreamdefaultreader.rs [hunk 6]
```diff
@@ -32,6 +34,19 @@ use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 pub enum ReadRequest {
     /// <https://streams.spec.whatwg.org/#default-reader-read>
     Read(Rc<Promise>),
+    /// Tee
+    Tee {
```
Comment (gterzian) [inline components/script/dom/readablestreamdefaultreader.rs]: Here we can also put the data in a `struct TeeReadRequest`.

File: components/script/dom/readablestreamdefaultreader.rs [hunk 7]
```diff
@@ -45,6 +60,83 @@ impl ReadRequest {
...
+                branch1,
+                branch2,
+                reading,
+                read_again,
+                canceled1,
+                canceled2,
+                clon_for_branch2,
+                cancel_promise: _,
+                global,
+            } => {
```
Comment (gterzian) [inline components/script/dom/readablestreamdefaultreader.rs]: Here for readability we should call into a method of the struct holding all the data, and document it with right link which here is `https://streams.spec.whatwg.org/#ref-for-read-request-chunk-steps%E2%91%A2`

File: components/script/dom/readablestreamdefaultreader.rs [hunk 8]
```diff
@@ -61,13 +153,64 @@ impl ReadRequest {
                     value: result,
                 });
             },
+            ReadRequest::Tee {
```
Comment (gterzian) [inline components/script/dom/readablestreamdefaultreader.rs]: same here with https://streams.spec.whatwg.org/#ref-for-read-request-close-steps%E2%91%A1

File: components/script/dom/readablestreamdefaultreader.rs [hunk 9]
```diff
@@ -61,13 +153,64 @@ impl ReadRequest {
...
+                }
+            },
         }
     }
 
     /// <https://streams.spec.whatwg.org/#ref-for-read-request-close-step>
     pub fn error_steps(&self, e: SafeHandleValue) {
         match self {
             ReadRequest::Read(promise) => promise.reject_native(&e),
+            ReadRequest::Tee {
```
Comment (gterzian) [inline components/script/dom/readablestreamdefaultreader.rs]: same here with https://streams.spec.whatwg.org/#ref-for-read-request-error-steps%E2%91%A2

File: components/script/dom/underlyingsourcecontainer.rs [hunk 10]
```diff
@@ -129,10 +151,65 @@ impl UnderlyingSourceContainer {
...
+                reading,
+                read_again,
+                canceled1,
+                canceled2,
+                clon_for_branch2,
+                reason1,
+                reason2,
+                cancel_promise,
+            } => {
+                // 13 - https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaulttee
```
Comment (gterzian) [inline components/script/dom/underlyingsourcecontainer.rs:259]: Here we can move it to a method of an inner struct as well, but unfortunately the spec has no specific link, so you can just document it as being the pull algorithm for either branch.

File: components/script/dom/underlyingsourcecontainer.rs [hunk 11]
```diff
@@ -34,6 +38,23 @@ pub enum UnderlyingSourceType {
...
+        stream: MutNullableDom<ReadableStream>,
+        branch1: MutNullableDom<ReadableStream>,
+        branch2: MutNullableDom<ReadableStream>,
+        reading: Rc<Cell<bool>>,
+        read_again: Rc<Cell<bool>>,
+        canceled1: Rc<Cell<bool>>,
+        canceled2: Rc<Cell<bool>>,
+        clon_for_branch2: Rc<Cell<bool>>,
+        #[no_trace]
+        reason1: RefCell<Option<JSVal>>,
```
Comment (gterzian) [inline components/script/dom/underlyingsourcecontainer.rs]: This should trace, I think you can use `Heap`.

**Review by gterzian — COMMENTED** [review]
Looks like it's going in the right direction. I think we should put the data of the enum variants into their own struct, and call methods on it, for readability and better docs.

**Pull request closed without merging**